pkg/queue/bounded_queue.go

```diff
@@ -21,89 +21,109 @@
 package queue
 
 import (
+	"sync"
+	"sync/atomic"
 	"time"
 
 	"github.com/uber/jaeger-lib/metrics"
 )
 
-// BoundedDropOldestQueue Implements a producer-consumer exchange similar to a ring buffer queue,
+// BoundedQueue implements a producer-consumer exchange similar to a ring buffer queue,
 // where the queue is bounded and if it fills up due to slow consumers, the new items written by
 // the producer force the earliest items to be dropped. The implementation is actually based on
 // channels, with a special Reaper goroutine that wakes up when the queue is full and consumers
 // the items from the top of the queue until its size drops back to maxSize
-type BoundedDropOldestQueue struct {
-	maxSize       int
+type BoundedQueue struct {
+	capacity      int
+	size          int32
 	onDroppedItem func(item interface{})
 	items         chan interface{}
-	reaps         chan interface{}
+	stopCh        chan struct{}
+	stopWG        sync.WaitGroup
+	stopped       int32
 }
 
-// MakeBoundedDropOldestQueue constructs the new queue of specified max size, and with an optional
-// callback for dropped items (e.g. useful to emit metrics)
-func MakeBoundedDropOldestQueue(maxSize int, overflow int, onDroppedItem func(item interface{})) *BoundedDropOldestQueue {
-	return &BoundedDropOldestQueue{
-		maxSize:       maxSize,
+// NewBoundedQueue constructs the new queue of specified capacity, and with an optional
+// callback for dropped items (e.g. useful to emit metrics).
+func NewBoundedQueue(capacity int, onDroppedItem func(item interface{})) *BoundedQueue {
+	return &BoundedQueue{
+		capacity:      capacity,
 		onDroppedItem: onDroppedItem,
-		items:         make(chan interface{}, maxSize+overflow),
-		reaps:         make(chan interface{}, overflow),
+		items:         make(chan interface{}, capacity),
+		stopCh:        make(chan struct{}),
 	}
 }
 
 // StartConsumers starts a given number of goroutines consuming items from the queue and passing them into
 // the consumer callback. Also starts the Reaper goroutine.
-func (q *BoundedDropOldestQueue) StartConsumers(num int, consumer func(item interface{})) {
+func (q *BoundedQueue) StartConsumers(num int, consumer func(item interface{})) {
 	for i := 0; i < num; i++ {
+		q.stopWG.Add(1)
 		go func() {
-			for item := range q.items {
-				consumer(item)
+			defer q.stopWG.Done()
+			for {
+				select {
+				case item := <-q.items:
+					atomic.AddInt32(&q.size, -1)
+					consumer(item)
+				case <-q.stopCh:
```

Comment:
I assume the order of `case` statements makes reading `q.items` a priority over `stopCh`, but don't have a definitive proof. If there is a priority, then this will implicitly try to drain the queue first. 

I don't think it matters for the span processor, in a real binary Stop never gets called anyway, it's only used in tests.

I did make more changes and completely refactored the test.